pkg/sfu/videolayerselector/framenumberwrapper.go

```diff
@@ -0,0 +1,41 @@
+package videolayerselector
+
+import "github.com/livekit/protocol/logger"
+
+type FrameNumberWrapper struct {
+	offset uint64
+	last   uint64
+	inited bool
+	logger logger.Logger
+}
+
+// UpdateAndGet returns the wrapped frame number from the given frame number, and updates the offset to
+// make sure the returned frame number is always inorder. Should only updateOffset if the new frame is a keyframe
+// because frame dependencies uses on the frame number diff so frames inside a GOP should have the same offset.
+func (f *FrameNumberWrapper) UpdateAndGet(new uint64, updateOffset bool) uint64 {
+	if !f.inited {
+		f.last = new
+		f.inited = true
+		return new
+	}
+
+	if new <= f.last {
+		return new + f.offset
+	}
+
+	if updateOffset {
+		new16 := uint16(new + f.offset)
+		last16 := uint16(f.last + f.offset)
+		// if new frame number wraps around and is considered as earlier by client, increase offset to make it later
+		if diff := new16 - last16; diff > 0x8000 || (diff == 0x8000 && new16 <= last16) {
+			// increase offset by 6000, nearly 10 seconds for 30fps video with 3 spatial layers
+			prevOffset := f.offset
+			f.offset += uint64(65535 - diff + 6000)
```

Comment:
The decoder doesn't expect the frame number to be contiguous since it is always not as we are always dropping frame which is not presented in the current target. It only cares about the frame number inorder,  this only happens when subscriber pause the track for a long time so I intentionally jump it for a distance to avoid the edge case that out of order packet arrived at this point (very rare, but might happen).